open-apparel-registry/open-apparel-registry
# Allow contributors to be verified

## Overview

Allow superusers to mark contributors as verified and show that verification status on the facility detail page and the contributor profile page.

Connects #545

## Demo

### Public Profile

<img width="1012" alt="Screen Shot 2019-06-06 at 1 40 18 PM" src="https://user-images.githubusercontent.com/17363/59065188-0050a080-8861-11e9-95a6-1ca2d8fc3856.png">

## My Profile Is Unchanged

<img width="1012" alt="Screen Shot 2019-06-06 at 1 40 32 PM" src="https://user-images.githubusercontent.com/17363/59065222-11011680-8861-11e9-8e3c-b92262a04902.png">

## Facility Details

<img width="1012" alt="Screen Shot 2019-06-06 at 1 40 58 PM" src="https://user-images.githubusercontent.com/17363/59065246-1c544200-8861-11e9-962c-9d447f60a7b0.png">

## Admin

<img width="882" alt="Screen Shot 2019-06-06 at 1 42 10 PM" src="https://user-images.githubusercontent.com/17363/59065259-24ac7d00-8861-11e9-9b27-c048e00a822f.png">

<img width="1099" alt="Screen Shot 2019-06-06 at 1 44 26 PM" src="https://user-images.githubusercontent.com/17363/59065301-3b52d400-8861-11e9-99a2-740d32521167.png">


## Notes

Adding a verified flag and a description box accessible via the Django Admin is the simplest possible version of this feature. We can grow a more complicated process from this if we discover a need. We opt the model into history tracking so that we have a log of changes to the verified status.

We have opted to show the name of the contributor as the title on the profile page rather than the static string "Profile" to emphasize what the page is about, to save some vertical space, and to provide a high-profile place on the page to show the verification badge.

We modified the `title` property on `AppGrid` to accept `node`s so we can include the icon component.

Showing the verified badge next to a contributor on the facility details page
requires us to have the verified status of contributors available in the
facility details API response. The previous implementation of the
`Facility.contributors` model method was returning a string representation of
the contributor. To better separate concerns, this helper now returns a list of
`FacilityList` objects that contributed the Facility, leaving the rendering up
to the serializer. We are now returning a list of objects that include an
`is_verified` flag that we use to conditionally render a verified SVG icon.

In order to reuse the same icon we used on the profile page, we are now passing
through the `style` prop from `BadgeVerified` to the internal `SvgIcon` in order
to adjust the size, margins, etc.

This PR also includes a unrelated change to the `Contributor` serializer that avoids multiple database calls. This was made before realizing that we would not need to change the `Contributor` serializer to implement the feature, but it is a useful change anyway.



## Testing Instructions

These instructions assume your database contains fixture data.

* Run migrations with `./scripts/manage migrate`
* Seed the new history data with `./scripts/manage populate_history --auto`
* Log in as a superuser, browse http://localhost:8081/admin/api/contributor/2/change/, and mark the contributor as verified.
* Browse http://localhost:6543/profile/2 and verify that the verified icon is displayed and that a "Verified" tooltip is shown on hover
* Browse http://localhost:6543/profile/3 and verify that no icon is displayed
* Browse http://localhost:6543/ and filter by contributor 2. Select a facility and verify that the verified icon is shown next to the contributor line on the detail page
* Filter by a different contributor and verify that there are no badges on the facility detail pages.
* Verify that the sample response on http://localhost:8081/api/docs/#!/facilities/facilities_read matches the actual API response shape.

## Checklist

- [x] Create a release issue and include a step for running `./scripts/manage populate_history --auto`
